File: src-tauri/src/tray_anim.rs
//! Animated menu-bar tray icon. The icon IS the brand mark — the dot-circle
//! "twinkle" (see src/renderer/components/LogoMark.tsx + sidebar/StatusWave.tsx):
//! a 7×7 disc of dots that breathe (scale 0.5→1, opacity 0.3→peak) on an eased,
//! desynced cycle. We rasterize that animation to a macOS *template* image
//! (black + alpha; the system tints it for the menu bar) and cycle frames,
//! driven by the aggregate session signal the transcript tailer computes:
//!
//!   BUSY      (any alive session "running")  → lively twinkle      ("working")
//!   YOUR_TURN (else any alive "waiting")     → slow unison pulse    ("your turn")
//!   IDLE      (no alive sessions)            → static mark
//!
//! The tailer writes the current state into `TrayState`; this module's thread
//! reads it ~12fps and repaints. `set_icon` marshals to the main thread itself.

use std::sync::atomic::{AtomicU8, Ordering};
use std::time::Duration;

use tauri::image::Image;
use tauri::Manager;

pub const IDLE: u8 = 0;
pub const BUSY: u8 = 1;
pub const YOUR_TURN: u8 = 2;

/// Aggregate tray signal — written by the transcript tailer, read by the animator.
#[derive(Default)]
pub struct TrayState(AtomicU8);

impl TrayState {
    pub fn set(&self, v: u8) {
        self.0.store(v, Ordering::Relaxed);
    }
    pub fn get(&self) -> u8 {
        self.0.load(Ordering::Relaxed)
    }
}

const SIZE: usize = 44; // matches icons/tray.png (22pt @2x)
const CENTER: f64 = 3.5; // grid centre in dot-centre coords: (7-1)/2 + 0.5
const DISC: f64 = 3.4; // disc radius in cell units (StatusWave RADIUS)
const DOT_R: f64 = 0.5; // dot radius in cell units
// Fit the 7-cell mark into a 40px disc with 2px padding so the rendered icon
// matches the visible size of the static tray.png (whose opaque box is 40×40 in
// the 44×44 canvas). Without the inset the disc filled the whole canvas and read
// slightly larger than the old icon.
const MARK: f64 = 40.0; // visible mark diameter in px
const INSET: f64 = (SIZE as f64 - MARK) / 2.0; // px padding on each side
const PX: f64 = MARK / 7.0; // pixels per cell (7 cells across the mark)

struct Dot {
    cx: f64,
    cy: f64,
    v: f64,   // size/brightness weight (LogoMark static)
    dur: f64, // breathing period (busy)
    off: f64, // phase offset (desync)
}

// Frozen pseudo-random matching the renderer's rand(): frac(sin(seed·k)·m).
fn rand(seed: f64) -> f64 {
    let x = (seed * 12.9898).sin() * 43758.5453;
    x - x.floor()
}

// Dots inside the disc, in the same c-outer/r-inner order as LogoMark/StatusWave
// so the per-index pseudo-random weights line up with the brand mark.
fn build_dots() -> Vec<Dot> {
    let max = DISC * DISC * 1.04;
    let mut out: Vec<Dot> = Vec::new();
    for c in 0..7 {
        for r in 0..7 {
            let cx = c as f64 + 0.5;
            let cy = r as f64 + 0.5;
            let dx = cx - CENTER;
            let dy = cy - CENTER;
            if dx * dx + dy * dy <= max {
                let i = out.len() as f64;
                out.push(Dot {
                    cx,
                    cy,
                    v: rand(i + 11.0),
                    dur: 1.4 + rand(i + 1.0) * 1.2, // 1.4–2.6s, the running tempo
                    off: rand(i + 7.0),
                });
            }
        }
    }
    out
}

// Anti-aliased filled disc into the alpha channel (black + alpha = template).
// Overlapping dots take the max alpha so seams don't double-darken.
fn paint_dot(buf: &mut [u8], cx: f64, cy: f64, scale: f64, op: f64) {
    let pcx = INSET + cx * PX;
    let pcy = INSET + cy * PX;
    let rad = scale * DOT_R * PX;
    let x0 = (pcx - rad - 1.0).floor().max(0.0) as usize;
    let x1 = (pcx + rad + 1.0).ceil().min(SIZE as f64) as usize;
    let y0 = (pcy - rad - 1.0).floor().max(0.0) as usize;
    let y1 = (pcy + rad + 1.0).ceil().min(SIZE as f64) as usize;
    for y in y0..y1 {
        for x in x0..x1 {
            let dx = x as f64 + 0.5 - pcx;
            let dy = y as f64 + 0.5 - pcy;
            let dist = (dx * dx + dy * dy).sqrt();
            let cov = (rad - dist + 0.5).clamp(0.0, 1.0); // ~1px AA edge
            if cov <= 0.0 {
                continue;
            }
            let a = (cov * op * 255.0) as u8;
            let idx = (y * SIZE + x) * 4 + 3; // alpha byte
            if a > buf[idx] {
                buf[idx] = a;
            }
        }
    }
}

// One RGBA frame for `state` at time `t` (seconds). RGB stays 0; only alpha
// carries the shape so macOS can tint it for the active menu bar.
fn frame(dots: &[Dot], state: u8, t: f64) -> Vec<u8> {
    let mut buf = vec![0u8; SIZE * SIZE * 4];
    // YOUR_TURN: the whole disc breathes in unison — a calm "your turn" beacon.
    let pulse = (1.0 - (std::f64::consts::TAU * (t / 2.6).fract()).cos()) * 0.5;
    for d in dots {
        let (op, scale) = match state {
            BUSY => {
                let phase = (t / d.dur + d.off).fract();
                let s = (1.0 - (std::f64::consts::TAU * phase).cos()) * 0.5; // 0→1→0
                (0.30 + 0.65 * s, 0.5 + 0.5 * s) // op 0.30→0.95, scale 0.5→1
            }
            YOUR_TURN => (0.35 + 0.40 * pulse, 0.62 + 0.30 * pulse), // gentle, legible
            _ => {
                let w = 0.62 + 0.38 * d.v; // IDLE: LogoMark's static weighting
                (w, w)
            }
        };
        paint_dot(&mut buf, d.cx, d.cy, scale, op);
    }
    buf
}

/// Spawn the repaint loop: read the aggregate signal ~12fps and update the tray.
pub fn start(app: tauri::AppHandle) {
    std::thread::spawn(move || {
        let dots = build_dots();
        let mut t = 0.0_f64;
        let mut idle_painted = false;
        // The your-turn pulse is a transient beacon, not a permanent state: after
        // YOUR_TURN_SETTLE seconds it settles to the static idle mark so the menu
        // bar doesn't pulse forever. Entering YOUR_TURN (incl. a fresh turn after
        // BUSY) re-arms it. Mirrors StatusWave's settle in the sidebar.
        const YOUR_TURN_SETTLE: f64 = 6.0;
        let mut prev_state = IDLE;
        let mut yt_elapsed = 0.0_f64;
        loop {
            std::thread::sleep(Duration::from_millis(80)); // ~12.5 fps
            let state = app.state::<TrayState>().get();
            if state == YOUR_TURN && prev_state != YOUR_TURN {
                yt_elapsed = 0.0; // a new your-turn stretch began → re-arm the pulse
            }
            prev_state = state;
            let Some(tray) = app.tray_by_id("operator") else {
                continue;
            };
            // A your-turn that has pulsed long enough renders as idle (static).
            let settled = state == YOUR_TURN && yt_elapsed >= YOUR_TURN_SETTLE;
            if state == IDLE || settled {
                // Static mark — paint once on settling, then rest the loop.
                if !idle_painted {
                    let buf = frame(&dots, IDLE, 0.0);
                    let _ = tray.set_icon_with_as_template(
                        Some(Image::new(&buf, SIZE as u32, SIZE as u32)),
                        true,
                    );
                    idle_painted = true;
                }
                continue;
            }
            idle_painted = false;
            if state == YOUR_TURN {
                yt_elapsed += 0.08;
            }
            t += 0.08;
            let buf = frame(&dots, state, t);
            let _ = tray
                .set_icon_with_as_template(Some(Image::new(&buf, SIZE as u32, SIZE as u32)), true);
        }
    });
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn rand_is_deterministic_and_in_unit_range() {
        assert_eq!(rand(0.0), 0.0); // sin(0)=0
        for i in 0..50 {
            let v = rand(i as f64 + 0.5);
            assert!((0.0..1.0).contains(&v), "rand out of range: {v}");
            assert_eq!(v, rand(i as f64 + 0.5)); // stable across calls
        }
    }

    #[test]
    fn build_dots_fills_the_disc_deterministically() {
        let dots = build_dots();
        // 7x7 grid clipped to the disc (radius 3.4, 1.04 tolerance) → 37 cells.
        assert_eq!(dots.len(), 37);
        let max = DISC * DISC * 1.04;
        for d in &dots {
            let dx = d.cx - CENTER;
            let dy = d.cy - CENTER;
            assert!(dx * dx + dy * dy <= max, "dot outside disc");
            assert!((1.4..2.6).contains(&d.dur), "dur out of band: {}", d.dur);
            assert!((0.0..1.0).contains(&d.v));
            assert!((0.0..1.0).contains(&d.off));
        }
    }
}

File: electron/src/main/tray-anim.ts
// The tray icon ANIMATES. Ported from `src-tauri/src/tray_anim.rs`.
//
// The icon IS the brand mark — the dot-circle "twinkle" (src/renderer/components/LogoMark.tsx,
// sidebar/StatusWave.tsx): a 7×7 disc of dots that breathe on an eased, desynced cycle. It is
// rasterized here to a macOS TEMPLATE image (black + alpha; the system tints it) and cycled from
// the aggregate lane signal:
//
//   busy      (any live lane "running")  → lively desynced twinkle   ("working")
//   your-turn (else any live "waiting")  → slow unison pulse         ("your turn")
//   idle      (neither)                  → the static mark, no timer
//
// MOTION IS THE BUSY SIGNAL — the house rule the sidebar follows too. Idle does not tick, it
// rests: a menu bar that animates forever is a menu bar you stop reading.
import { nativeImage } from 'electron'

export type TrayPhase = 'idle' | 'busy' | 'your-turn'

/** The aggregate signal, ported from the reducer the tailer runs inline
 *  (src-tauri/src/transcript.rs:1131-1149). Anything working outranks anything waiting: the
 *  twinkle means "Operator is doing something", and one running lane makes that true. */
export function aggregateState(lanes: Array<{ phase: string }>): TrayPhase {
  let waiting = false
  for (const l of lanes) {
    if (l.phase === 'running') return 'busy'
    if (l.phase === 'waiting') waiting = true
  }
  return waiting ? 'your-turn' : 'idle'
}

const SIZE = 44 // the canvas: 22pt @2x, the size a macOS tray image is handed in
const CENTER = 3.5 // grid centre in dot-centre coords: (7-1)/2 + 0.5
const DISC = 3.4 // disc radius in cell units (StatusWave RADIUS)
const DOT_R = 0.5 // dot radius in cell units

/** THE VISIBLE MARK — 36px at scaleFactor 2, i.e. **18pt**.
 *
 *  18pt is what a macOS menu-bar template icon is: the system's own status items draw at roughly
 *  that, and anything larger reads as an app that did not measure. This was 40px (20pt), matched
 *  to the opaque box inside `tray.png` — which was itself the thing that was too big. Measured
 *  before the change: ink spanned x 2..41 in every phase, 40px across, 20.0pt.
 *
 *  The CANVAS stays 44. Shrinking the canvas instead would have been the same pixels with a
 *  different container and would have moved the problem to whatever else assumed 44. */
const MARK = 36.0 // visible mark diameter in px → 18pt at scaleFactor 2
const INSET = (SIZE - MARK) / 2 // 4px padding on each side
const PX = MARK / 7 // pixels per cell (7 cells across the mark)
const TAU = Math.PI * 2

interface Dot {
  cx: number
  cy: number
  v: number // size/brightness weight (LogoMark static)
  dur: number // breathing period (busy)
  off: number // phase offset (desync)
}

/** Frozen pseudo-random matching the renderer's `rand()`: frac(sin(seed·k)·m). The point is not
 *  randomness, it is that the tray's dots carry the SAME weights as the mark in the app. */
export function rand(seed: number): number {
  const x = Math.sin(seed * 12.9898) * 43758.5453
  return x - Math.floor(x)
}

/** Dots inside the disc, in the same c-outer/r-inner order as LogoMark/StatusWave so the
 *  per-index weights line up with the brand mark. */
export function buildDots(): Dot[] {
  const max = DISC * DISC * 1.04
  const out: Dot[] = []
  for (let c = 0; c < 7; c++) {
    for (let r = 0; r < 7; r++) {
      const cx = c + 0.5
      const cy = r + 0.5
      const dx = cx - CENTER
      const dy = cy - CENTER
      if (dx * dx + dy * dy <= max) {
        const i = out.length
        out.push({
          cx,
          cy,
          v: rand(i + 11),
          dur: 1.4 + rand(i + 1) * 1.2, // 1.4-2.6s, the running tempo
          off: rand(i + 7),
        })
      }
    }
  }
  return out
}

/** Anti-aliased filled disc into the ALPHA channel only. Overlapping dots take the max alpha so
 *  seams don't double-darken. */
function paintDot(buf: Buffer, cx: number, cy: number, scale: number, op: number): void {
  const pcx = INSET + cx * PX
  const pcy = INSET + cy * PX
  const rad = scale * DOT_R * PX
  const x0 = Math.max(0, Math.floor(pcx - rad - 1))
  const x1 = Math.min(SIZE, Math.ceil(pcx + rad + 1))
  const y0 = Math.max(0, Math.floor(pcy - rad - 1))
  const y1 = Math.min(SIZE, Math.ceil(pcy + rad + 1))
  for (let y = y0; y < y1; y++) {
    for (let x = x0; x < x1; x++) {
      const dx = x + 0.5 - pcx
      const dy = y + 0.5 - pcy
      const dist = Math.sqrt(dx * dx + dy * dy)
      const cov = Math.min(1, Math.max(0, rad - dist + 0.5)) // ~1px AA edge
      if (cov <= 0) continue
      const a = Math.trunc(cov * op * 255)
      const idx = (y * SIZE + x) * 4 + 3 // alpha byte
      if (a > buf[idx]) buf[idx] = a
    }
  }
}

/** One RGBA frame for `state` at time `t` (seconds).
 *
 *  RGB STAYS ZERO - only alpha carries the shape, which is what makes it a template image macOS
 *  can tint for the active menu bar, and incidentally why the byte order does not matter: the
 *  raw-bitmap format is platform-dependent (BGRA on most), but with all three colour bytes at 0
 *  RGBA and BGRA are the same bytes, premultiplied or not. */
export function frame(dots: Dot[], state: TrayPhase, t: number): Buffer {
  const buf = Buffer.alloc(SIZE * SIZE * 4)
  // your-turn: the whole disc breathes in unison - a calm "your turn" beacon.
  const p = t / 2.6
  const pulse = (1 - Math.cos(TAU * (p - Math.floor(p)))) * 0.5
  for (const d of dots) {
    let op: number
    let scale: number
    if (state === 'busy') {
      const ph = t / d.dur + d.off
      const s = (1 - Math.cos(TAU * (ph - Math.floor(ph)))) * 0.5 // 0->1->0
      op = 0.3 + 0.65 * s // 0.30->0.95
      scale = 0.5 + 0.5 * s // 0.5->1
    } else if (state === 'your-turn') {
      op = 0.35 + 0.4 * pulse // gentle, legible
      scale = 0.62 + 0.3 * pulse
    } else {
      const w = 0.62 + 0.38 * d.v // idle: LogoMark's static weighting
      op = w
      scale = w
    }
    paintDot(buf, d.cx, d.cy, scale, op)
  }
  return buf
}

/** A frame as a template `NativeImage`, ready for `tray.setImage`. */
export function frameImage(dots: Dot[], state: TrayPhase, t: number): Electron.NativeImage {
  const img = nativeImage.createFromBitmap(frame(dots, state, t), { width: SIZE, height: SIZE, scaleFactor: 2 })
  img.setTemplateImage(true)
  return img
}

const TICK_MS = 80 // ~12.5fps - enough for a breathing dot, cheap enough to ignore
/** The your-turn pulse is a transient beacon, not a permanent state: after this it settles to
 *  the static mark so the menu bar doesn't pulse forever. Entering your-turn again (including a
 *  fresh turn after busy) re-arms it. Mirrors StatusWave's settle in the sidebar. */
const YOUR_TURN_SETTLE_S = 6

/** Drive the icon: cadence, the settle, and the decision NOT to paint. Returns the stop function.
 *
 *  It hands `paint` a state and a time rather than an image, which keeps every rule about WHEN
 *  the tray repaints testable under plain node — rasterizing needs Electron, timing does not.
 *
 *  The state is PULLED rather than pushed so a missed event cannot leave the icon stuck
 *  mid-twinkle: every tick asks what is true now. `paint` is only called when there is something
 *  new to show - idle paints once and then the loop rests. */
export function startTrayAnimation(
  getState: () => TrayPhase,
  paint: (state: TrayPhase, t: number) => void,
): () => void {
  let t = 0
  let idlePainted = false
  let prev: TrayPhase = 'idle'
  let ytElapsed = 0
  const tick = () => {
    const state = getState()
    if (state === 'your-turn' && prev !== 'your-turn') ytElapsed = 0 // a new stretch -> re-arm
    prev = state
    const settled = state === 'your-turn' && ytElapsed >= YOUR_TURN_SETTLE_S
    if (state === 'idle' || settled) {
      if (!idlePainted) {
        paint('idle', 0)
        idlePainted = true
      }
      return
    }
    idlePainted = false
    if (state === 'your-turn') ytElapsed += TICK_MS / 1000
    t += TICK_MS / 1000
    paint(state, t)
  }
  tick() // paint the static mark immediately rather than 80ms into the app's life
  const timer = setInterval(tick, TICK_MS)
  return () => clearInterval(timer)
}
